python/ray/data/impl/plan.py

```diff
@@ -91,27 +136,106 @@ def execute(self, clear_input_blocks: bool = True) -> BlockList:
             self._out_stats.dataset_uuid = self._dataset_uuid
         return self._out_blocks
 
-    def clear(self) -> None:
-        self._out_blocks = None
-        self._out_stats = None
+    def _optimize(self) -> None:
+        """Apply stage fusion optimizations, updating this plan."""
+        context = DatasetContext.get_current()
+        if context.optimize_fuse_stages:
+            if context.optimize_fuse_read_stages:
+                self._rewrite_read_stages()
+            self._fuse_one_to_one_stages()
+
+    def _rewrite_read_stages(self) -> None:
+        """Rewrites read stages into one-to-one stages."""
+        if self._stages and self._has_read_stage():
+            block_list, stage = self._rewrite_read_stage()
+            self._in_blocks = block_list
+            self._in_stats = DatasetStats(stages={}, parent=None)
+            self._stages.insert(0, stage)
+
+    def _has_read_stage(self) -> bool:
+        """Whether this plan has a read stage for its input."""
+        return isinstance(self._in_blocks, LazyBlockList) and hasattr(
+            self._in_blocks, "_read_tasks"
+        )
+
+    def _is_read_stage(self) -> bool:
+        """Whether this plan is a bare read stage."""
+        return self._has_read_stage() and not self._stages
+
+    def _rewrite_read_stage(self) -> Tuple[BlockList, "Stage"]:
+        """Rewrite the read stage to a OneToOne stage over read tasks as input.
+
+        For example, suppose the plan was [Read -> MapBatches(Fn)]. These stages cannot
+        be fused, since read stages are handled specially.
+
+        After rewriting to [GetReadTasks -> MapBatches(DoRead) -> MapBatches(Fn)],
+        now we can fuse the latter two MapBatches stages into a single OneToOne stage:
+        [GetReadTasks -> MapBatches(DoRead -> Fn)].
+        """
+        # Generate the "GetReadTasks" stage blocks.
+        blocks = []
+        metadata = []
+        for i, read_task in enumerate(self._in_blocks._read_tasks):
+            blocks.append(ray.put([read_task]))
```

Comment:
Nit/future PR: It would be nice if we could enable this fusion without having to put the read tasks into the object store since read tasks can have non-negligible serialization overhead (see issues with Parquet dataset piece pickling taking a long time).

Reply:
I don't think this is a notable issue. The read tasks are already sent across the network anyways in the tasks, so this isn't really changing much.

Reply:
Hmm I suppose that the Parquet dataset pieces are already pre-serialized here. Good point!